flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/QueryOperationConverter.java

```diff
@@ -130,8 +138,39 @@ public RelNode visit(QueryOperation other) {
 		}
 
 		@Override
-		public <U> RelNode visit(TableSourceQueryOperation<U> tableSourceTable) {
-			throw new UnsupportedOperationException("Unsupported now");
+		public <U> RelNode visit(TableSourceQueryOperation<U> tableSourceOperation) {
+			TableSource<?> tableSource = tableSourceOperation.getTableSource();
+			boolean isBatch = tableSourceOperation.isBatch();
+			FlinkStatistic statistic;
+			List<String> names;
+			if (tableSourceOperation instanceof RichTableSourceQueryOperation &&
+				((RichTableSourceQueryOperation<U>) tableSourceOperation).getQualifiedName() != null) {
+				statistic = ((RichTableSourceQueryOperation<U>) tableSourceOperation).getStatistic();
+				names = ((RichTableSourceQueryOperation<U>) tableSourceOperation).getQualifiedName();
+			} else {
+				statistic = FlinkStatistic.UNKNOWN();
+				// TableSourceScan requires a unique name of a Table for computing a digest.
+				// We are using the identity hash of the TableSource object.
+				String refId = "Unregistered_TableSource_" + System.identityHashCode(tableSource);
+				names = Collections.singletonList(refId);
+			}
+
+			TableSourceTable<?> tableSourceTable = new TableSourceTable<>(tableSource, !isBatch, statistic);
+			FlinkRelOptTable table = FlinkRelOptTable.create(
+				relBuilder.getRelOptSchema(),
+				tableSourceTable.getRowType(relBuilder.getTypeFactory()),
+				names,
+				tableSourceTable);
+
+			// when LogicalTableScan converts a table scan with QueryOperationCatalogViewTable to a new table scan
+			// with TableSourceTable, both two table scans have same digest due to both two tables have
+			// exactly the same table name. so use EnumerableTableScan instead of LogicalTableScan to
+			// make sure two table scans have different digest.
+			if (createEnumerableTableScan) {
```

Comment:
If I understand this correctly this is the problem of introducing the `qualifiedName` in `RichTableSourceQueryOperation`, right? Can we not introduce hacks in production code to make the tests work?

Reply:
As above.